File: crates/ldw/src-rs/ldw/model/analysed/in_place_transformer.rs
// Generated on 2024-11-13T12:00:22.739Z
use super::model::*;

pub trait InPlaceTransformer {
    #[allow(unused_mut)]
    fn transform_model(self: &mut Self, mut node: Box<Model>) -> Box<Model> {
        node.name = self.transform_fqn(node.name);
        node.parent = node.parent.map(|value| self.transform_model(value));
        node.definitions = node
            .definitions
            .into_iter()
            .map(|(key, value)| (key, self.transform_definition(value)))
            .collect();
        node
    }

    #[allow(unused_mut)]
    fn transform_fqn(self: &mut Self, mut node: Fqn) -> Fqn {
        node.into_iter().map(|value| value).collect()
    }

    #[allow(unused_mut)]
    fn transform_definition(self: &mut Self, mut node: Box<Definition>) -> Box<Definition> {
        node.r#type = self.transform_type(node.r#type);
        node
    }

    #[allow(unused_mut)]
    fn transform_type(self: &mut Self, mut node: Type) -> Type {
        match node {
            Type::VoidType(value) => Type::VoidType(self.transform_void_type(value)),
            Type::PrimitiveType(value) => Type::PrimitiveType(self.transform_primitive_type(value)),
            Type::EnumType(value) => Type::EnumType(self.transform_enum_type(value)),
            Type::TypeWithStructure(value) => {
                Type::TypeWithStructure(self.transform_type_with_structure(value))
            }
            Type::NamedTypeReference(value) => {
                Type::NamedTypeReference(self.transform_named_type_reference(value))
            }
        }
    }

    #[allow(unused_mut)]
    fn transform_void_type(self: &mut Self, mut node: Box<VoidType>) -> Box<VoidType> {
        node
    }

    #[allow(unused_mut)]
    fn transform_primitive_type(self: &mut Self, mut node: PrimitiveType) -> PrimitiveType {
        node
    }

    #[allow(unused_mut)]
    fn transform_enum_type(self: &mut Self, mut node: Box<EnumType>) -> Box<EnumType> {
        node
    }

    #[allow(unused_mut)]
    fn transform_type_with_structure(
        self: &mut Self,
        mut node: TypeWithStructure,
    ) -> TypeWithStructure {
        match node {
            TypeWithStructure::SumType(value) => {
                TypeWithStructure::SumType(self.transform_sum_type(value))
            }
            TypeWithStructure::ProductType(value) => {
                TypeWithStructure::ProductType(self.transform_product_type(value))
            }
            TypeWithStructure::GenericType(value) => {
                TypeWithStructure::GenericType(self.transform_generic_type(value))
            }
        }
    }

    #[allow(unused_mut)]
    fn transform_sum_type(self: &mut Self, mut node: Box<SumType>) -> Box<SumType> {
        node.members = node
            .members
            .into_iter()
            .map(|value| self.transform_type(value))
            .collect();
        node
    }

    #[allow(unused_mut)]
    fn transform_product_type(self: &mut Self, mut node: Box<ProductType>) -> Box<ProductType> {
        node.members = node
            .members
            .into_iter()
            .map(|value| self.transform_product_member(value))
            .collect();
        node
    }

    #[allow(unused_mut)]
    fn transform_product_member(
        self: &mut Self,
        mut node: Box<ProductMember>,
    ) -> Box<ProductMember> {
        node.r#type = self.transform_type(node.r#type);
        node
    }

    #[allow(unused_mut)]
    fn transform_generic_type(self: &mut Self, mut node: GenericType) -> GenericType {
        match node {
            GenericType::MapType(value) => GenericType::MapType(self.transform_map_type(value)),
            GenericType::SetType(value) => GenericType::SetType(self.transform_set_type(value)),
            GenericType::SequenceType(value) => {
                GenericType::SequenceType(self.transform_sequence_type(value))
            }
            GenericType::OptionType(value) => {
                GenericType::OptionType(self.transform_option_type(value))
            }
        }
    }

    #[allow(unused_mut)]
    fn transform_map_type(self: &mut Self, mut node: Box<MapType>) -> Box<MapType> {
        node.key_type = self.transform_type(node.key_type);
        node.value_type = self.transform_type(node.value_type);
        node
    }

    #[allow(unused_mut)]
    fn transform_set_type(self: &mut Self, mut node: Box<SetType>) -> Box<SetType> {
        node.key_type = self.transform_type(node.key_type);
        node
    }

    #[allow(unused_mut)]
    fn transform_sequence_type(self: &mut Self, mut node: Box<SequenceType>) -> Box<SequenceType> {
        node.element_type = self.transform_type(node.element_type);
        node
    }

    #[allow(unused_mut)]
    fn transform_option_type(self: &mut Self, mut node: Box<OptionType>) -> Box<OptionType> {
        node.r#type = self.transform_type(node.r#type);
        node
    }

    #[allow(unused_mut)]
    fn transform_named_type_reference(
        self: &mut Self,
        mut node: Box<NamedTypeReference>,
    ) -> Box<NamedTypeReference> {
        node.fqn = self.transform_fqn(node.fqn);
        node
    }
}

File: crates/codegen/ldw/src-ts/languages/ldw/model/resolved/inPlaceTransformer.ts
import * as Model from './model';

export class InPlaceTransformer {
    transformModel(input: Model.Model): Model.Model {
        throw new Error('Automatic transformation not possible');
    }

    transformDefinition(input: Model.Definition): Model.Definition {
        return new Model.Definition({ name: input.name, type: this.transformTypeToType(input.type) });
    }

    // Each `...ToType` form is per-context in which the type is found

    transformTypeToType(input: Model.Type): Model.Type {
        if (input instanceof Model.VoidType) {
            return this.transformVoidTypeToType(input);
        } else if (input instanceof Model.PrimitiveType) {
            return this.transformPrimitiveTypeToType(input);
        } else if (input instanceof Model.EnumType) {
            return this.transformEnumTypeToType(input);
        } else if (input instanceof Model.NamedTypeReference) {
            return this.transformNamedTypeReferenceToType(input);
        } else if (input instanceof Model.SumType) {
            return this.transformSumTypeToType(input);
        } else if (input instanceof Model.ProductType) {
            return this.transformProductTypeToType(input);
        } else if (input instanceof Model.MapType) {
            return this.transformMapTypeToType(input);
        } else if (input instanceof Model.SetType) {
            return this.transformSetTypeToType(input);
        } else if (input instanceof Model.SequenceType) {
            return this.transformSequenceTypeToType(input);
        } else {
            return this.transformOptionTypeToType(input);
        }
    }

    // When this is generated, it should follow the type hierarchy and include the following:
    // transformdTypeWithStructureToType(input: Model.StructType): Model.Type
    // transformGenericTypeToType(input: Model.GenericType): Model.Type

    transformVoidTypeToType(input: Model.VoidType): Model.Type {
        return input;
    }

    transformPrimitiveTypeToType(input: Model.PrimitiveType): Model.Type {
        return input;
    }

    transformEnumTypeToType(input: Model.EnumType): Model.Type {
        return input;
    }

    transformNamedTypeReferenceToType(input: Model.NamedTypeReference): Model.Type {
        return input;
    }

    transformSumTypeToType(input: Model.SumType): Model.Type {
        return new Model.SumType({ members: input.members.map((m) => this.transformTypeToType(m)) });
    }

    transformProductTypeToType(input: Model.ProductType): Model.Type {
        return new Model.ProductType({ members: input.members.map((m) => this.transformProductMember(m)) });
    }

    transformProductMember(input: Model.ProductMember): Model.ProductMember {
        return new Model.ProductMember({ name: input.name, type: this.transformTypeToType(input.type) });
    }

    transformMapTypeToType(input: Model.MapType): Model.Type {
        return new Model.MapType({
            keyType: this.transformTypeToType(input.keyType),
            valueType: this.transformTypeToType(input.valueType)
        });
    }

    transformSetTypeToType(input: Model.SetType): Model.Type {
        return new Model.SetType({ keyType: this.transformTypeToType(input.keyType) });
    }

    transformSequenceTypeToType(input: Model.SequenceType): Model.Type {
        return new Model.SequenceType({ elementType: this.transformTypeToType(input.elementType) });
    }

    transformOptionTypeToType(input: Model.OptionType): Model.Type {
        return new Model.OptionType({ type: this.transformTypeToType(input.type) });
    }
}
